apis/Google.Cloud.AspNetCore.DataProtection.SecretManager/Google.Cloud.AspNetCore.DataProtection.SecretManager/CloudSecretManagerXmlRepository.cs

```diff
@@ -0,0 +1,128 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Xml.Linq;
+using Secret = Google.Cloud.SecretManager.V1.Secret;
+
+namespace Google.Cloud.AspNetCore.DataProtection.SecretManager;
+
+/// <summary>
+/// Implementation of <see cref="IXmlRepository"/> that stores the protected elements in a Google Cloud Secret Manager.
+/// This class is configured by <see cref="GoogleCloudDataProtectionBuilderExtensions.PersistKeysToGoogleCloudSecretManager(Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder, string, string)"/>
+/// (and other overloads).
+/// </summary>
+internal class CloudSecretManagerXmlRepository : IXmlRepository
+{
+    private readonly SecretManagerServiceClient _secretManagerServiceClient;
+    private readonly SecretName _secretName;
+    private readonly ProjectName _projectName;
+    private readonly SecretVersionName _version;
+    private readonly CallSettings _callSettings = CallSettingsExtensions.WithRetry(CallSettings.FromExpiration(Expiration.FromTimeout(TimeSpan.FromMilliseconds(10000))), RetrySettings.FromExponentialBackoff(maxAttempts: 5, initialBackoff: TimeSpan.FromMilliseconds(0.2), maxBackoff: TimeSpan.FromMilliseconds(10000), backoffMultiplier: 1.5, retryFilter: RetrySettings.FilterForStatusCodes(StatusCode.DeadlineExceeded, StatusCode.Unavailable)));
+
+    internal CloudSecretManagerXmlRepository(SecretManagerServiceClient client, string secretName, string projectName)
+    {
+        _secretManagerServiceClient = GaxPreconditions.CheckNotNull(client, nameof(client));
+        GaxPreconditions.CheckNotNull(secretName, nameof(secretName));
+        GaxPreconditions.CheckNotNull(projectName, nameof(projectName));
+        _projectName = new ProjectName(projectName);
+        _secretName = new SecretName(projectName, secretName);
+
+        // Version is populated in such a way that it always refers to the most latest version.
+        // For more information on version naming refer to
+        // https://cloud.google.com/dotnet/docs/reference/Google.Cloud.SecretManager.V1/latest/Google.Cloud.SecretManager.V1.AccessSecretVersionRequest#Google_Cloud_SecretManager_V1_AccessSecretVersionRequest_Name
+        _version = SecretVersionName.FromProjectSecretSecretVersion(projectName, secretName, "latest");
+    }
+
+    /// <inheritdoc />
+    public IReadOnlyCollection<XElement> GetAllElements()
+    {
+        if (!SecretExists())
+        {
+            return Array.Empty<XElement>();
+        }
+        var accessSecretVersionResponse = GetSecretVersionResponse();
+
+        if (accessSecretVersionResponse != null)
+        {
+            var secretPayload = accessSecretVersionResponse.Payload.Data.ToStringUtf8();
+            XDocument document = XDocument.Parse(secretPayload);
+            var xElements = document?.Root?.Elements().ToList();
+            return xElements ?? (IReadOnlyCollection<XElement>) Array.Empty<XElement>();
```

Comment:
I think .ToList() will never return null.

Reply:
ToList won't, but if `document` or `document.Root` is null (which is explicitly handled in the line above) then `xElements` will still be null. That said, I don't believe `XDocument.Parse` ever returns null, or returns a document with no root.

Reply:
got it I will address it.